In [1]:
import tensorflow as tf
import keras
from keras.datasets import mnist
from keras.models import Sequential
from keras.layers import Dense, Dropout, Activation, Flatten
from keras.layers import Conv2D, MaxPooling2D
from keras.utils import np_utils
import numpy as np


np.random.seed(42)

In [2]:
(x_train, y_train), (x_test, y_test) = mnist.load_data()

x_train = x_train.reshape(60000, 784)
x_test = x_test.reshape(10000, 784)

x_train = x_train.astype("float32")
x_test = x_test.astype("float32")

x_train /= 255
x_test /= 255

y_train = np_utils.to_categorical(y_train, 10)
y_test = np_utils.to_categorical(y_test, 10)

print(x_train.shape, 'x_train')
print(y_train.shape, 'y_train')
print(x_test.shape, 'x_test')
print(y_test.shape, 'y_test')

11493376/11490434 [==============================] - 2s 0us/step
(60000, 784) x_train
(60000, 10) y_train
(10000, 784) x_test
(10000, 10) y_test


In [3]:
# полносвязная сеть
model = Sequential()
model.add(Dense(units=100, input_dim=784, activation='relu'))
model.add(Dense(units=10, activation='softmax'))

model.summary()
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])
model.fit(x_train, y_train, batch_size=128, epochs=5, verbose=2, validation_data=(x_test, y_test))
score,accuracy = model.evaluate(x_test, y_test,verbose=0)

print('Test accuracy:', accuracy)

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense (Dense)                (None, 100)               78500     
_________________________________________________________________
dense_1 (Dense)              (None, 10)                1010      
Total params: 79,510
Trainable params: 79,510
Non-trainable params: 0
_________________________________________________________________
Epoch 1/5
469/469 - 1s - loss: 0.3780 - accuracy: 0.8971 - val_loss: 0.2101 - val_accuracy: 0.9414
Epoch 2/5
469/469 - 1s - loss: 0.1755 - accuracy: 0.9503 - val_loss: 0.1533 - val_accuracy: 0.9550
Epoch 3/5
469/469 - 1s - loss: 0.1287 - accuracy: 0.9635 - val_loss: 0.1219 - val_accuracy: 0.9633
Epoch 4/5
469/469 - 1s - loss: 0.1026 - accuracy: 0.9707 - val_loss: 0.1024 - val_accuracy: 0.9699
Epoch 5/5
469/469 - 1s - loss: 0.0840 - accuracy: 0.9748 - val_loss: 0.0917 - val_accuracy: 0.9731
Test accuracy: 0.97

In [4]:
(x_train, y_train), (x_test, y_test) = mnist.load_data()

x_train = x_train.reshape(x_train.shape[0], 28, 28, 1)
x_test = x_test.reshape(x_test.shape[0], 28, 28, 1)

x_train = x_train.astype('float32')
x_test = x_test.astype('float32')

x_train /= 255
x_test /= 255

y_train = np_utils.to_categorical(y_train, 10)
y_test = np_utils.to_categorical(y_test, 10)

print(x_train.shape, 'x_train')
print(y_train.shape, 'y_train')
print(x_test.shape, 'x_test')
print(y_test.shape, 'y_test')

(60000, 28, 28, 1) x_train
(60000, 10) y_train
(10000, 28, 28, 1) x_test
(10000, 10) y_test


In [6]:
# сверточная сеть
model = Sequential()

model.add(Conv2D(32, (3, 3), padding='valid', input_shape=(28, 28, 1)))
model.add(Activation('relu'))
model.add(Conv2D(32, (3, 3), activation ='relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.25))
model.add(Flatten())
model.add(Dense(128))
model.add(Activation('relu'))
model.add(Dropout(0.5))
model.add(Dense(10))
model.add(Activation('softmax'))

In [7]:
model.summary()
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])
model.fit(x_train, y_train, batch_size=128, epochs=5, verbose=2, validation_data=(x_test, y_test))
score,accuracy = model.evaluate(x_test, y_test,verbose=0)

print('Test accuracy:', accuracy)

Model: "sequential_2"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_2 (Conv2D)            (None, 26, 26, 32)        320       
_________________________________________________________________
activation_3 (Activation)    (None, 26, 26, 32)        0         
_________________________________________________________________
conv2d_3 (Conv2D)            (None, 24, 24, 32)        9248      
_________________________________________________________________
max_pooling2d_1 (MaxPooling2 (None, 12, 12, 32)        0         
_________________________________________________________________
dropout_2 (Dropout)          (None, 12, 12, 32)        0         
_________________________________________________________________
flatten_1 (Flatten)          (None, 4608)              0         
_________________________________________________________________
dense_4 (Dense)              (None, 128)              

В целом полносвязная сеть хорошо справляется с датасетом MNIST - точность на уровне 97%, но сверточная сеть при существенном конечно приросте времени на обучение справляется еще лучше - почти 99% точности на тестовом наборе.